source/simulator/stokes_matrix_free.cc

```diff
@@ -304,6 +371,436 @@ namespace aspect
                 }
       }
 
+      template <int dim, int spacedim>
+      void compute_no_normal_flux_constraints_on_level(const DoFHandler<dim, spacedim> &dof_handler,
+                                                       const MGConstrainedDoFs &mg_constrained_dofs,
+                                                       const Mapping<dim> &mapping,
+                                                       const unsigned int level,
+                                                       const unsigned int first_vector_component,
+                                                       const std::set<types::boundary_id> &boundary_ids,
+                                                       AffineConstraints<double> &constraints)
+      {
+        // TODO: This is a simplification of compute_no_normal_flux_constraints() from deal.II.
+        // The differences are:
+        // - It works on a specific level so we can ignore hanging nodes
+        // - We use the normal vector given by the manifold (instead of averaging surface vectors)
+        //
+        // This should go into deal.II at some point, but it is too specific right now.
+
+        // three cases:
+        // 1. one manifold normal vector for each vector dofs, compute_no_normal_flux_constraints_shell
+        // 2. dim manifold normal vector for each vector dofs, set each components to zero
+        // 3. dim -1 = 2 manifold normal vector for each vector dofs (dim = 3), v*n1 = 0, v*n2 = 0
+
```

Comment:
I think you need to handle 1 vector here as well. If you have all sides of a 3d chunk as no flux, you have a mixture of all three cases.